etc/keepalived/kp_local_loop.py

```diff
@@ -0,0 +1,693 @@
+                # service down
+                # Master/Slave Switch
+                switch_process(node_obj)
+                if not node_obj.alarmed:
+                    notify_issue(node_obj, affect_services=service_obj,
+                                 affect_range='node',
+                                 more_specific='necessary_error')
+                    service_obj.post_alarmed_if_possible()
+                    node_obj.post_alarmed_if_possible()
+            else:
+                # Have report the issue yet?
+                # Yes
+                if service_obj.alarmed:
+
+                    # --- the issue is timeout ?
+                    # -- Yes
+                    if service_obj.is_alarmed_timeout():
+
+                        # $$ this node is master ?
+                        # $$ Yes
+                        # Report an issue towards deployment as a unnecessary
+                        # service issue timeout.
+                        # Master/Slave Switch
+                        if node_obj.role == 'master':
+                            switch_process(node_obj)
+                            if not node_obj.alarmed:
+                                notify_issue(node_obj, affect_range='node',
+                                             more_specific='slave_switch')
+                                node_obj.post_alarmed_if_possible()
+                    # -- No
+                    # report node heartbeat
+                    else:
+                        # We can just return/go through,
+                        # as the outside loop will set the node heartbeat
+                        pass
+
+                # No
+                # Report an issue towards node and service as a
+                # unnecessary service down.
+                # Then report node heartbeat.
+                else:
+                    if not service_obj.alarmed:
+                        notify_issue(node_obj, affect_services=service_obj,
+                                     affect_range='service',
+                                     more_specific='unecessary_svc')
+                        service_obj.post_alarmed_if_possible()
+        else:
+            # (TODO) rsync, gearman, zuul-timer-tasks,
+            # nodepool-timer-tasks are not
+            # support to restart.
+            if service_obj.name not in ['rsync', 'gearman', 'zuul-timer-tasks',
+                                        'nodepool-timer-tasks']:
+                # restart the specific service
+                service_restart(SYSTEMCTL_RESTART, service_obj.name)
+                service_obj.post_restarted_if_possible()
+
+
+def shut_down_all_services(node_obj):
+    for service in node_obj.services:
+        run_systemctl_command(SYSTEMCTL_STOP, service.name)
+    node_obj.services = []
+
+
+def setup_necessary_services_and_check(node_obj):
+    for service in node_obj.necessary_service_names:
+        run_systemctl_command(SYSTEMCTL_START, service)
+
```

Comment:
```
for service in node_obj.services:
    if service['is_necessary']:
        run_systemctl_command(SYSTEMCTL_START, service)
```

Then the `necessary_service_names` property can be removed totally.

Reply:
done